### PR TITLE
fix(engine): directory targets sanity check

### PR DESCRIPTION
**Pull request by rgrinberg**
The current check is wrong because it assumes that all directory targets
are going to be produced for the current directory. Some directories
forward their rules to parents, which means that sub-directories should
only bother checking for targets that are declared for them.

It's not possible to reproduce this test using rules defined in `dune` files at the moment. It's only possible to reproduce this by defining rules using the `Dune_engine` api.

Another option is to skip this check altogether when invoking the rules through a sub-dir (a bit trickier to do).

### CODE REVIEW

**Comment by snowleopard** [comment]
The new check seems too weak to me. A better check should clearly depend on whether the targets are actually redirected or not.  Also, I don't like that we can't test it. 

I'd rather we don't support directory targets when they are redirected at all instead of weakening the check.

**Comment by rgrinberg** [comment]
> The new check seems too weak to me. A better check should clearly depend on whether the targets are actually redirected or not

Yes, I agree. Let me improve the check.

> Also, I don't like that we can't test it.

This PR came out of https://github.com/ocaml/dune/pull/5695. Where we used directory targets to generate documentation rules for coq. We can easily show that the check fixes the issue there. It's not ideal of course, and I do think that we should unit test the public api of dune_engine as well. Alas, we're not ready for that yet.

> I'd rather we don't support directory targets when they are redirected at all instead of weakening the current check.

That would be a shame because the feature is very convenient for doc targets.

**Comment by rgrinberg** [comment]
@snowleopard I put a new version of the sanity check. It appears to be correct in all cases and the validation error is raised as soon as possible. As a bonus, I improved the error message to guide the developer a little better. Previously, the error left me wondering what was wrong, the new one is easier to understand imo

File: src/dune_engine/load_rules.ml [hunk 1]
```diff
@@ -630,6 +630,35 @@ end = struct
             && Subdir_set.mem build_dir_only_sub_dirs name
           then report_rule_internal_dir_conflict name loc);
       let* rules_produced = Memo.Lazy.force rules in
+      let () =
+        let real_directory_targets = Rules.directory_targets rules_produced in
+        if
+          not
+            (Path.Build.Map.equal real_directory_targets directory_targets
+               ~equal:(fun _ _ ->
+                 (* the locations don't matter and aren't expected to match *)
```
[the remark this answers is not in the document]
Comment (snowleopard) [reply src/dune_engine/load_rules.ml:639]: Oh, I see. Yes, I think this needs clarifying. Not only in this comment, but also in the comments in `Rules.directory_targets` and `Gen_rules`?

**Review by snowleopard — APPROVED** [review]
Thanks, looks good now (modulo the `Loc.t` confusion).

I'm a bit puzzled though: it's not obvious that this change actually changes behaviour. I suppose `Rules.directory_targets` somehow manages to be more accurate than `rules_here.by_directory_targets`. Why is that?

**Comment by rgrinberg** [comment]
rules_here.by_directory_targets is the directory targets for the directory where we are loading the rules. While Rules.directory_targets is for all produced rules which includes subdirectories.

The current check fails because it compares a declaration for all produced rules vs the directory targets in the dir we are loading.

My first attempt at the fix was to narrow the declaration to only the subset that matters to the dir we are loading.

This new fix checks the full declaration against the full set of directory targets from the produced rules.

**Comment by snowleopard** [comment]
Got it, thanks!

**Pull request merged**